src/core/cardano/walletConnect/identityWalletConnect.ts

```diff
@@ -45,15 +53,41 @@ class IdentityWalletConnect extends CardanoPeerConnect {
 
     this.sign = async (
       identifier: string,
-      payload: string
+      payload: string,
+      onSignError?: (error: Error) => void
     ): Promise<string> => {
-      if (this.signerCache.get(identifier) === undefined) {
-        this.signerCache.set(
+      let approved: boolean | undefined = undefined;
+      // Closure that updates approved variable
+      const approvalCallback = (approvalStatus: boolean) => {
+        approved = approvalStatus;
+      };
+      this.eventService.emit<PeerConnectSigningEvent>({
+        type: PeerConnectSigningEventTypes.PeerConnectSign,
+        payload: {
           identifier,
-          await Agent.agent.identifiers.getSigner(identifier)
-        );
+          payload,
+          approvalCallback,
+        },
+      });
+      // Wait until approved is true or false
+      while (approved === undefined) {
+        await new Promise((resolve) => setTimeout(resolve, 1000));
+      }
+      if (approved) {
+        if (this.signerCache.get(identifier) === undefined) {
+          this.signerCache.set(
+            identifier,
+            await Agent.agent.identifiers.getSigner(identifier)
+          );
+        }
+        return this.signerCache.get(identifier)!.sign(Buffer.from(payload))
+          .qb64;
+      } else {
+        if (typeof onSignError === "function") {
+          onSignError(new Error("User declined the signing request"));
```

Comment:
There is a specific error type from CIP-30. Do we have access to import this?

`DataSignError` and `DataSignErrorCode.UserDeclined`

Reply:
Yeah, I saw the error types in the CIP-30. However, I don't know how to import those types. It looks like we need to define it ourselves. Btw, the `onSignError` that Jaime told me about receives an `Error` as a parameter, and `DataSignErrorCode.UserDeclined` is only a number.

Reply:
`DataSignError` is the error - `DataSignErrorCode.UserDeclined` is just an attribute of the error.

```
DataSignErrorCode {
	ProofGeneration: 1,
	AddressNotPK: 2,
	UserDeclined: 3,
}
type DataSignError = {
	code: DataSignErrorCode,
	info: String
}
```

At the very least, we need to follow this format. If there's nothing to import them from, lets just define them ourselves.